### PR TITLE
Anko - Update DataSource

### PR DESCRIPTION
**Pull request by at-toannguyen1**
#### Related issues (Required)
- Task #132 
- Bug #

#### Description (Required)
- [ ] Update Datasource

### CODE REVIEW

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 1]
```diff
@@ -13,47 +20,141 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
  */
 internal class WayRemoteDataSource : WayDataSource {
     override fun createUser(userParams: UserParams): Observable<ResponseStatus> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        val result = AsyncSubject.create<ResponseStatus>()
+
+
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 2]
```diff
@@ -13,47 +20,141 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
 
     override fun searchLocations(input: String, key: String, language: String, sensor: Boolean)
             : Observable<AutoCompleteResult> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return ApiClient.instance.searchLocations(input, key, language, sensor).toObservable()
     }
 
     override fun createGroup(name: String): Observable<Group> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return HypertrackApi.instance.createGroup(name).toObservable()
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 3]
```diff
@@ -13,47 +20,141 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
     }
 
     override fun getGroupInfo(groupId: String): Observable<Group> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return HypertrackApi.instance.getGroupInfo(groupId).toObservable()
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 4]
```diff
@@ -13,47 +20,141 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
     }
 
     override fun getGroupMembers(groupId: String): Observable<MutableList<User>> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return HypertrackApi.instance.getGroupMembers(groupId).toObservable().map { it.results }
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 5]
```diff
@@ -13,47 +20,141 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
     }
 
     override fun addUserToGroup(userId: String, body: BodyAddUserToGroup): Observable<User> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return HypertrackApi.instance.addUserToGroup(userId, body).toObservable()
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 6]
```diff
@@ -13,47 +20,141 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
+        return result
     }
 
-    override fun getUser(userId: String): Observable<User> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+    override fun getUser(): Observable<User> {
+        val result = AsyncSubject.create<User>()
+        HyperTrack.getUser(object : HyperTrackCallback() {
+            override fun onSuccess(response: SuccessResponse) {
+                val res = response.responseObject as User
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: ERROR: Cast operator throws an exception if the cast is not possible.

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 7]
```diff
@@ -13,47 +20,141 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
+                val err = Throwable(error.errorMessage)
+                result.onError(err)
+            }
+        })
+        return result
+    }
+
+    override fun getETA(destination: LatLng, vehicle: VehicleType): Observable<Float> {
+        val result = AsyncSubject.create<Float>()
+        HyperTrack.getETA(destination, vehicle, object : HyperTrackCallback() {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: ERROR: Cast operator throws an exception if the cast is not possible.

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 8]
```diff
@@ -13,47 +20,141 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
+                result.onError(err)
+            }
+
+        })
+        return result
+    }
+
+    override fun createAndAssignAction(builder: ActionParamsBuilder): Observable<Action> {
+        val result = AsyncSubject.create<Action>()
+        HyperTrack.createAndAssignAction(builder.build(), object : HyperTrackCallback() {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: ERROR: Cast operator throws an exception if the cast is not possible.

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 9]
```diff
@@ -13,47 +20,141 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
+                val res = Throwable(error.errorMessage)
+                result.onError(res)
+            }
+        })
+        return result
     }
 
     override fun getAddressLocation(latLng: String): Observable<MutableList<LocationAddress>> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return ApiClient.instance.getAddressLocation(latLng).toObservable().map { it.results }
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt (hunk 2 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt (hunk 3 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt (hunk 4 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt (hunk 5 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt [hunk 10]
```diff
@@ -1,58 +1,73 @@
...
     }
 
     override fun updateUser(userParams: UserParams): Observable<ResponseStatus> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return remoteDataSource.updateUser(userParams)
     }
 
-    override fun getUser(userId: String): Observable<User> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+    override fun getUser(): Observable<User> {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt [hunk 11]
```diff
@@ -1,58 +1,73 @@
...
     }
 
     override fun getLocationDetail(placeId: String?, key: String): Observable<ResultPlaceDetail> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return remoteDataSource.getLocationDetail(placeId, key)
     }
 
     override fun searchLocations(input: String, key: String, language: String, sensor: Boolean): Observable<AutoCompleteResult> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return remoteDataSource.searchLocations(input, key, language, sensor)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt [hunk 12]
```diff
@@ -1,58 +1,73 @@
...
     }
 
     override fun addUserToGroup(userId: String, body: BodyAddUserToGroup): Observable<User> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return remoteDataSource.addUserToGroup(userId, body)
     }
 
     override fun removeUserFromGroup(userId: String, body: BodyAddUserToGroup): Observable<User> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return remoteDataSource.removeUserFromGroup(userId, body)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt [hunk 13]
```diff
@@ -1,58 +1,73 @@
...
+    }
+
+    override fun getCurrentLocation(): Observable<HyperTrackLocation> {
+        return remoteDataSource.getCurrentLocation()
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt [hunk 14]
```diff
@@ -1,58 +1,73 @@
...
  * Copyright © 2017 AsianTech inc.
  * Created by vinh.huynh on 11/24/17.
  */
-class WayRepository : WayDataSource{
+class WayRepository : WayDataSource {
+    private val remoteDataSource = WayRemoteDataSource()
+
     override fun createUser(userParams: UserParams): Observable<ResponseStatus> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return remoteDataSource.createUser(userParams)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt (hunk 12 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt (hunk 13 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 15]
```diff
@@ -13,47 +20,137 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
 
     override fun searchLocations(input: String, key: String, language: String, sensor: Boolean)
             : Observable<AutoCompleteResult> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return ApiClient.instance.searchLocations(input, key, language, sensor).toObservable()
     }
 
     override fun createGroup(name: String): Observable<Group> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return HypertrackApi.instance.createGroup(name).toObservable()
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 16]
```diff
@@ -13,47 +20,137 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
     }
 
     override fun getGroupInfo(groupId: String): Observable<Group> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return HypertrackApi.instance.getGroupInfo(groupId).toObservable()
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 17]
```diff
@@ -13,47 +20,137 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
     }
 
     override fun getGroupMembers(groupId: String): Observable<MutableList<User>> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return HypertrackApi.instance.getGroupMembers(groupId).toObservable().map { it.results }
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 18]
```diff
@@ -13,47 +20,137 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
     }
 
     override fun addUserToGroup(userId: String, body: BodyAddUserToGroup): Observable<User> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return HypertrackApi.instance.addUserToGroup(userId, body).toObservable()
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 19]
```diff
@@ -13,47 +20,137 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
+                val res = Throwable(error.errorMessage)
+                result.onError(res)
+            }
+        })
+        return result
     }
 
     override fun getAddressLocation(latLng: String): Observable<MutableList<LocationAddress>> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        return ApiClient.instance.getAddressLocation(latLng).toObservable().map { it.results }
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt (hunk 15 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt (hunk 16 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt (hunk 17 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt (hunk 18 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt (hunk 10 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt (hunk 11 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt (hunk 12 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt (hunk 13 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt (hunk 14 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt (hunk 12 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt (hunk 13 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/WayRepository.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt [hunk 20]
```diff
@@ -13,47 +20,137 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
+        val result = AsyncSubject.create<ResponseStatus>()
+        HyperTrack.getOrCreateUser(userParams, object : HyperTrackCallback() {
+            override fun onSuccess(response: SuccessResponse) {
+                val res = ResponseStatus(true, "Success")
+                result.onNext(res)
+                result.onComplete()
+            }
+
+            override fun onError(error: ErrorResponse) {
+                val res = Throwable(error.errorMessage)
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/way/data/source/remote/WayRemoteDataSource.kt]: `res ` change to `throwable`, also check on this file

**Comment by at-vinhhuynh** [comment]
@at-tienhoang 
I finish review this PR, please check it again.
thank you!

**Pull request merged**